### PR TITLE
Idefics 3 support

### PR DESCRIPTION
**Pull request by pcuenca**
I didn't find anyone working on it yet, so decided to give it a go.

### CODE REVIEW

File: mlx_vlm/models/idefics3/vision.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        super().__init__()
+        self.activation_fn = FastGELUActivation()
+        self.fc1 = nn.Linear(config.hidden_size, config.intermediate_size, bias=True)
+        self.fc2 = nn.Linear(config.intermediate_size, config.hidden_size, bias=True)
+
+    def __call__(self, x: mx.array) -> mx.array:
+        x = self.fc1(x)
+        x = self.activation_fn(x)
+        x = self.fc2(x)
+        return x
```
Comment (Blaizzy) [inline mlx_vlm/models/idefics3/vision.py]: Nit:  MLX has this [Gelu activation](https://ml-explore.github.io/mlx/build/html/python/nn/_autosummary/mlx.nn.GELU.html#mlx.nn.GELU). 
```suggestion
class MLP(nn.Module):
    def __init__(self, config: VisionConfig):
        super().__init__()
        self.activation_fn = nn.GELU(approx='precise')
        self.fc1 = nn.Linear(config.hidden_size, config.intermediate_size, bias=True)
        self.fc2 = nn.Linear(config.intermediate_size, config.hidden_size, bias=True)

    def __call__(self, x: mx.array) -> mx.array:
        x = self.fc1(x)
        x = self.activation_fn(x)
        x = self.fc2(x)
        return x
```

**Review by Blaizzy — CHANGES_REQUESTED** [review]
Thank you very much for your contribution @pcuenca!

Idefics 3 is an awesome model that has been on my radar.

Just 3 tiny changes before we merge:
1. `pre-commit run --all-files` (to fix style tests)
2. Change GELU to default 
3. Add model tests [here](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/tests/test_models.py)

**Comment by Blaizzy** [comment]
My pleasure!

I noticed, and am very happy to see you contributing. It means a lot to me! 🎉❤️

File: mlx_vlm/models/idefics3/vision.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        )
+
+        self.num_patches = (self.image_size // self.patch_size) ** 2
+        self.num_positions = self.num_patches
+        self.position_embedding = nn.Embedding(self.num_positions, self.embed_dim)
+
+    def __call__(self, x: mx.array) -> mx.array:
+        patch_embeddings = self.patch_embedding(x)
+        patch_embeddings = mx.flatten(patch_embeddings, start_axis=1, end_axis=2)
+        position_ids = mx.array(np.arange(self.num_positions)[None, :])
```
Comment (Blaizzy) [inline mlx_vlm/models/idefics3/vision.py]: Nit: 

MLX has [mx.arange](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.arange.html#mlx.core.arange) built-in.

Mixing numpy/torch in the graph execution leads to significant performance slow down.

Read more [here](https://github.com/Blaizzy/mlx-vlm/issues/110#issuecomment-2474931063).

```suggestion
        position_ids = mx.array(mx.arange(self.num_positions)[None, :])
```

**Review by Blaizzy — CHANGES_REQUESTED** [review]

File: mlx_vlm/models/idefics3/vision.py (hunk 2 above)
Comment (Blaizzy) [reply mlx_vlm/models/idefics3/vision.py]: This change and the GELU might get you a few extra tokens-per-sec (+3).

File: mlx_vlm/models/idefics3/language.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        inputs: mx.array,
+        cache=None,
+        inputs_embeds=None,
+        mask: Optional[mx.array] = None,
+    ):
+        # for passing merged input embeddings
+        if inputs_embeds is None:
+            h = self.embed_tokens(inputs)
+        else:
+            h = inputs_embeds
```
Comment (awni) [inline mlx_vlm/models/idefics3/language.py]: I recommend setting the `inputs_embeds` to the model dtype here. It will be faster since you will avoid unexpectedly upcasting weights to a higher precision.
```suggestion
            h = inputs_embeds.astype(self.norm.weight.dtype)
```

Comment (awni) [reply mlx_vlm/models/idefics3/language.py]: (With this change + fixing the GeLU I get a nice speedup)

**Comment by awni** [comment]
I was playing with this model locally and it works quite well (and is even faster with some of the suggested changes).

One odd thing I noticed is the vision encoder is quantized with type `float32` whereas the LM is quantized with `float16`. I traced it to the padding operation which I think inadvertently [always uses `float32`](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/utils.py#L634-L635). Those `mx.zeros` will default to `float32`. You can fix that by doing:

```python
    new_weight = mx.zeros((new_out, new_in), weight.dtype)
    new_bias = mx.zeros((new_out), weight.dtype)
```

File: mlx_vlm/models/idefics3/vision.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        self.encoder = Encoder(config)
+        self.post_layernorm = nn.LayerNorm(config.hidden_size)
+
+    def __call__(
+        self,
+        x: mx.array,
+        output_hidden_states: Optional[bool] = None,
+    ) -> mx.array:
+        x = self.embeddings(x)
+        encoder_outputs = self.encoder(
```
Comment (awni) [inline mlx_vlm/models/idefics3/vision.py:204]: Probably also good to cast `x` to the model type here. Right now it doesn't matter much since the model type is `fp32` due to what I think is a bug in the quantization step. But if that's fixed this will matter a lot (will use less RAM and avoid upcasting the model).

```suggestion
        x = x.astype(self.embeddings.patch_embedding.weight.dtype)
        encoder_outputs = self.encoder(
```

**Comment by Blaizzy** [comment]
Thanks a lot @awni!  

I will fix it as soon as I get back home.

File: mlx_vlm/models/idefics3/vision.py (hunk 2 above)
Comment (pcuenca) [reply mlx_vlm/models/idefics3/vision.py]: oh nice, this was an oversight, my bad!

**Comment by pcuenca** [comment]
~114 tok/s with your fixes @Blaizzy @awni 🤯 🙌

**Comment by Blaizzy** [comment]
Awesome, that’s the power of pure MLX 🚀😎

I’m about to merge the utils, then rebase this PR and merge.

**Comment by Blaizzy** [comment]
@pcuenca have you uploaded the models to the HUB?

If not, where do you want the the weights to live? 

Is it ok to put them in the MLX-Community repo?

**Comment by pcuenca** [comment]
> Is it ok to put them in the MLX-Community repo?

Of course!

**Comment by Blaizzy** [comment]
Thanks, uploading the models right now :)
